### ISSUE TITLE
Reconstruction hangs on certain events

### ISSUE THREAD

**Issue by mholtrop** [JeffersonLab/hps-java#267]
We still have events getting stuck for recon. 

So far, the two distinct cases I spotted:

1. Recon gets stuck in a loop, and there is no output anymore, but CPU continues to be consumed. It is truely stuck on a single event. This is the case for hps_005783.evio.105
1. We get “stuck" in a loop that occasionally still spits something out (hps_005796.evio.212 does this). However, it appears recon is not truly stuck,it is just in a funk, taking forever (as in 3 to 4 hours) to process a single event.  Here is the console message that comes out
2017-12-06 08:54:14 [WARNING] org.hps.recon.tracking.TrackerReconDriver process :: 5796 53569438 Discarding track with bad HelicalTrackHit (correction distance 0.000000, chisq penalty 0.000000)

Additionally, somewhat obscure behavior: in an evio file that tries to isolate the event, in this case event 27207416 from run 5783 file 105, by copying the evio file but skipping the first 222545 events, so I start recon on event 27207411, the code DOES NOT get stuck. If however I start on event 27207410 or earlier, the code DOES get stuck.

**Comment by mdiamon** [comment]
Regarding the console message that comes out in case (2), I went down the rabbit-hole of this message awhile ago.
org.lcsim.fit.helicaltrack.HelicalTrackCross.setTrackDirection (which corrects HelicalTrackHits for parallax due to track direction) has an `errok` check: https://github.com/slaclab/lcsim/blob/master/tracking/src/main/java/org/lcsim/fit/helicaltrack/HelicalTrackCross.java#L141 
if this check fails, it calls resetTrackDirection, which puts the hit's correction and chi-squared to 0 instead of setting them properly.  then the message gets spit out.
the errok check relies partly on the uncorrected covariance matrix, which we already know is constructed in a lousy way (the entries are too large by a factor of a few.)  it also relies on a parameter  _eps that seems to be set arbitrarily.  that being said ... the check only fails when the track direction is nearly parallel to the sensor plane.  presumably this is a weird track, so it kind of makes sense that it would somehow be stumping the recon.

**Comment by mdiamon** [comment]
Regarding the obscure behavior that seems to depend on what event you start at: Norman reported a similar behavior in #243 .  I fixed the root cause of the runtime exception that was sometimes being thrown ... but didn't investigate why it was sometimes thrown and sometimes wasn't.  I'll keep burrowing down that rabbit-hole; perhaps I'll find an explanation.

**Comment by mholtrop** [comment]
I found the reason why, in this case, the events were processed differently depending on where you start in the file. We have a TriggerBank with a time stamp only once every 10 trigger banks. This timestamp is used to lookup in the conditions database whether the svt_bias was set correctly. So if the first few events don't have the time stamp, then these are skipped because the svt_bias is set to bad. After the first timestamp the processing continues as normal. In other words, the svt_bias setting has a resolution of one in every 10 events, which probably is fine.

**Comment by mholtrop** [comment]
I found one of the issues which causes runs to get stuck is too many hits i the SVT. These high noise events cause an explosion in the number of iterations in SeedTrackFinder, which then goes off to never never land. *It should not*!!. Even with very large iterations, it should eventually come out, but this seems to not actually happen, at least, not within my 10 hours of testing time. 

The fix which is now in RawTrackerHitFitter is to abort the event if there are more than N hits. Sebouh checked one file from the 2015 dataset and another from the 2016 dataset, and got these results:   requiring nRawHits <= 300 would cut out about ~0.3% of the total “good” trident events.  (both tracks have fit chi^2<50, n_sigma < 10, unconstrained vertex chi^2 < 100), and positron has L1 hit and d0< 1.1 mm).   If we want a little looser, nRawHits <= 400 would cut out 0.04% of the events , and at 500, <0.01% would be cut from the event.

It seems that setting the cut at 500, or a 1000 is quite safe and will also save us some processing time.

Not all the files Rafo flagged completed. I will run them again, but will continue digging deeper as well.

**Comment by normangraf** [comment]
Thanks for tracking this down Maurik. This could, in principle, cause issues when processing files containing only selected events, but I think if we only select good events we should be OK.

Good work,

Norman


________________________________

**Comment by normangraf** [comment]
Thanks Maurik,


The number of hits grows much faster than linearly, since we get a profusion of ghost hits when making the 3D hits from axial and stereo combinations.


I would opt for the higher numbers (500 or 1000) and see how that works.


Good work,

Norman


________________________________

**Comment by sebouh137** [comment]
I have set up branch iss271 to set a maximum number of hits.  Currently,
the default (in the driver) is 300, but this can be modified via changes to
the steering file.  Maurik, if you want me to set the default maximum
number of hits to 500 or 1000 in there (or change it in the steering
files), I can certainly do so.

**Comment by mholtrop** [comment]
Hi Sebouh,

I am not sure why we needed this branch 271 and issue 271, since branch iss267 is already dealing with these large hit number events for issue 267. In that branch the event is aborted and a warning is printed using the logger. The default value for the maximum is set at a super conservative 2000, which is just low enough to not get stuck on any events. That value can be set in the options to the driver. Also, if the threshold is crossed, the entire event should be aborted, just as it would when the svt_bias_good flag is false.

I am currently writing a script to analyze the log files of a number of data files to see the correlation between number of hits, number of iteration in SeedTrackFinder and total event time. Which I hope will give some guidance on where to set the max hits.

Best,
	Maurik

**Comment by sebouh137** [comment]
I was unaware of branch iss267 when I created branch iss271.  I will be
more careful in the future to check if branches/issues already exist before
creating new ones.

**Comment by mholtrop** [comment]
Seems this issue is closed, as we have not had reports of additional hangs of reconstruction. If the problem re-occurs, we can open a new issue.
